raiden/raiden_event_handler.py

```diff
@@ -295,13 +299,40 @@ def handle_contract_send_channelunlock(
             raiden: RaidenService,
             channel_unlock_event: ContractSendChannelBatchUnlock,
     ):
-        channel = raiden.chain.payment_channel(
+        payment_channel: PaymentChannel = raiden.chain.payment_channel(
             channel_unlock_event.token_network_identifier,
             channel_unlock_event.channel_identifier,
         )
+        token_network: TokenNetwork = payment_channel.token_network
+
+        # Fetch on-chain balance hashes for both participants
+        participants_details = token_network.detail_participants(
+            raiden.address,
+            channel_unlock_event.partner,
+            channel_unlock_event.channel_identifier,
+        )
+
+        # Replay state changes until a channel state is reached where
+        # this channel state has the participants balance hash.
+        restored_channel_state = channel_state_until_balance_hash(
+            raiden=raiden,
+            token_address=channel_unlock_event.token_address,
+            channel_identifier=channel_unlock_event.channel_identifier,
+            target_balance_hash=participants_details.partner_details.balance_hash,
+        )
+
+        if not restored_channel_state:
+            log.error('Could not find channel state with partner balance hash {}'.format(
```

Comment:
This is a bug and IMO should be an unrecoverable error

Reply:
+1 good point. The user should be immediately alerted of this and send us a bug report.